# Exploring a Used Car Price Dataset

## Notebook goal

In this notebook, we explore the used car dataset before cleaning the data and training regression models.

The goal of this analysis is to understand:

- what information the dataset contains;
- how the vehicle characteristics are structured;
- what types of values appear in each column;
- whether missing or inconsistent values exist;
- which problems need to be solved during the cleaning phase.

The target variable in this project is:

`priceUSD`

which represents the vehicle price in US dollars.

We will not clean the data in this notebook.

We will only inspect the dataset and create a clear checklist of tasks that need to be handled before building the machine learning model.

## 1. Introduction

Before training a regression model, it is necessary to understand the dataset and identify possible problems in the raw data.

A used car dataset can contain different issues that may negatively affect machine learning performance:

- inconsistent column names;
- missing values;
- incorrect data types;
- duplicated records;
- unrealistic vehicle prices;
- extreme mileage values;
- inconsistent categorical values.

If these problems are not detected before training, the model may learn from incorrect information and produce unreliable predictions.

The purpose of this notebook is to perform exploratory data analysis (EDA) and identify the main data quality issues before the cleaning and preprocessing stages.

## 2. Import libraries

In this section, we import the libraries required for exploratory data analysis.

The following libraries will be used:

- `pandas` - for loading, exploring, and manipulating tabular data;
- `numpy` - for numerical operations;
- `matplotlib` - for creating basic visualizations;
- `seaborn` - for statistical visualizations and better-looking plots.

These libraries will help us understand the structure, quality, and characteristics of the used car dataset.

In [33]:
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


## 3. Define the data path

The dataset used in this project is stored in the `data` folder.

The project structure is organized as follows:

car-price-prediction/

├── data/

│   └── cars.csv

└── notebooks/

    └── 01_eda.ipynb


The path below allows the notebook to access the dataset from the correct project location.

If the dataset location changes, only this path needs to be updated.

In [34]:
DATA_PATH = Path("../data/cars.csv")

DATA_PATH

WindowsPath('../data/cars.csv')

## 4. Load the dataset

The dataset used in this project contains information about used cars collected from vehicle advertisements.

Each row represents one used car advertisement, while each column describes a specific characteristic of the vehicle.

The dataset contains the following information:

- `make` - vehicle brand;
- `model` - vehicle model;
- `priceUSD` - vehicle price in US dollars (target variable);
- `year` - production year of the vehicle;
- `condition` - vehicle condition;
- `mileage(kilometers)` - total vehicle mileage in kilometers;
- `fuel_type` - type of fuel used by the vehicle;
- `volume(cm3)` - engine volume in cubic centimeters;
- `color` - vehicle color;
- `transmission` - type of transmission;
- `drive_unit` - type of drivetrain;
- `segment` - vehicle segment.

The target variable in this project is:

`priceUSD`

The goal of the regression model is to predict the expected market value of a used vehicle based on its characteristics.

At this stage, we only load and inspect the raw dataset. Data cleaning, feature engineering, and preprocessing will be performed in later stages of the project.

In [35]:
df = pd.read_csv(DATA_PATH)


## 5. First look at the data

After loading the dataset, we perform an initial inspection of the records.

The purpose of this step is to understand how the data is organized and whether the values appear in the expected format.

We will inspect:

- the first rows of the dataset;
- a random sample of records.

A random sample is useful because the first rows may not always represent the whole dataset.

In [36]:
df.head()

,make,model,priceUSD,year,condition,mileage(kilometers),fuel_type,volume(cm3),color,transmission,drive_unit,segment
0,mazda,2,5500,2008,with mileage,162000.0,petrol,1500.0,burgundy,mechanics,front-wheel drive,B
1,mazda,2,5350,2009,with mileage,120000.0,petrol,1300.0,black,mechanics,front-wheel drive,B
2,mazda,2,7000,2009,with mileage,61000.0,petrol,1500.0,silver,auto,front-wheel drive,B
3,mazda,2,3300,2003,with mileage,265000.0,diesel,1400.0,white,mechanics,front-wheel drive,B
4,mazda,2,5200,2008,with mileage,97183.0,diesel,1400.0,gray,mechanics,front-wheel drive,B


The first rows give us a quick overview of the dataset structure.

We can verify:

- whether all expected columns are present;
- whether values are loaded correctly;
- whether there are obvious formatting issues.

In [37]:
df.sample(5, random_state=42)

,make,model,priceUSD,year,condition,mileage(kilometers),fuel_type,volume(cm3),color,transmission,drive_unit,segment
19270,mitsubishi,carisma,2050,1999,with mileage,250000.0,petrol,1800.0,blue,mechanics,front-wheel drive,M
25927,ford,fusion,4500,2006,with mileage,160000.0,petrol,1400.0,burgundy,mechanics,front-wheel drive,M
23388,mercedes-benz,e-klass,3500,1999,with mileage,485000.0,diesel,2900.0,blue,mechanics,rear drive,E
53189,bmw,x3,21000,2013,with mileage,159000.0,diesel,2000.0,black,auto,NaN,J
34058,renault,megane,3990,2002,with mileage,331700.0,diesel,1900.0,other,mechanics,front-wheel drive,C


### Initial observations

The initial inspection confirms that the dataset contains vehicle information that can be used as input features for a regression model.

Further exploration is required to understand:

- dataset dimensions;
- data types;
- missing values;
- numerical and categorical features;
- possible outliers and invalid values.

These observations will guide the data cleaning and preprocessing steps.

## 6. Basic dataset structure

After the initial inspection, we analyze the basic structure of the dataset.

Understanding the dataset dimensions and data types helps us identify possible issues before the cleaning phase.

In this section, we will inspect:

- number of rows and columns;
- column names;
- detected data types;
- number of non-missing values.

In [38]:
df.shape

(56244, 12)

In [39]:
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")

Number of rows: 56244
Number of columns: 12


- Column names:



In [40]:
df.columns.tolist()

['make',
 'model',
 'priceUSD',
 'year',
 'condition',
 'mileage(kilometers)',
 'fuel_type',
 'volume(cm3)',
 'color',
 'transmission',
 'drive_unit',
 'segment']

### Dataset information

The `info()` method provides an overview of:

- column names;
- detected data types;
- number of non-null values;
- memory usage.

This information helps us identify columns that may require type conversion or missing value handling.

In [41]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 56244 entries, 0 to 56243
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   make                 56244 non-null  str    
 1   model                56244 non-null  str    
 2   priceUSD             56244 non-null  int64  
 3   year                 56244 non-null  int64  
 4   condition            56244 non-null  str    
 5   mileage(kilometers)  56244 non-null  float64
 6   fuel_type            56244 non-null  str    
 7   volume(cm3)          56197 non-null  float64
 8   color                56244 non-null  str    
 9   transmission         56244 non-null  str    
 10  drive_unit           54339 non-null  str    
 11  segment              50953 non-null  str    
dtypes: float64(2), int64(2), str(8)
memory usage: 5.1 MB


## 7. Understanding the meaning of columns

Before cleaning the data, it is important to understand the role and meaning of each column.

The dataset contains different types of information about vehicles.

The columns can be grouped into:

- vehicle identification information;
- numerical vehicle characteristics;
- categorical vehicle characteristics;
- target variable.

This grouping helps us decide how each type of data should be processed during the preprocessing stage.

### Vehicle identification columns

These columns describe the basic identity of the vehicle.

They contain information about the manufacturer and the specific vehicle model.

In [42]:
vehicle_identification_columns = [
    "make",
    "model"
]

df[vehicle_identification_columns].head()

,make,model
0,mazda,2
1,mazda,2
2,mazda,2
3,mazda,2
4,mazda,2


### Numerical columns

Numerical features represent measurable vehicle characteristics.

These columns can be used as numerical inputs for regression models after appropriate preprocessing.

In [43]:
numerical_columns = [
    "priceUSD",
    "year",
    "mileage(kilometers)",
    "volume(cm3)"
]

df[numerical_columns].head()

,priceUSD,year,mileage(kilometers),volume(cm3)
0,5500,2008,162000.0,1500.0
1,5350,2009,120000.0,1300.0
2,7000,2009,61000.0,1500.0
3,3300,2003,265000.0,1400.0
4,5200,2008,97183.0,1400.0


### Categorical columns

Categorical features describe vehicle characteristics that represent different groups or categories.

These columns will require encoding before they can be used by machine learning algorithms.

In [44]:
categorical_columns = [
    "make",
    "model",
    "condition",
    "fuel_type",
    "color",
    "transmission",
    "drive_unit",
    "segment"
]

df[categorical_columns].head()

,make,model,condition,fuel_type,color,transmission,drive_unit,segment
0,mazda,2,with mileage,petrol,burgundy,mechanics,front-wheel drive,B
1,mazda,2,with mileage,petrol,black,mechanics,front-wheel drive,B
2,mazda,2,with mileage,petrol,silver,auto,front-wheel drive,B
3,mazda,2,with mileage,diesel,white,mechanics,front-wheel drive,B
4,mazda,2,with mileage,diesel,gray,mechanics,front-wheel drive,B


### Target variable

The target variable is the value that the regression model will predict.

In this project, the target variable is:

`priceUSD`

It represents the market price of the used vehicle in US dollars.

In [45]:
target_column = "priceUSD"

df[[target_column]].head()

,priceUSD
0,5500
1,5350
2,7000
3,3300
4,5200


## 8. Target column inspection

The target variable is the value that the regression model will learn to predict.

In this project, the target variable is:

`priceUSD`

which represents the used car price in US dollars.

Before training the model, we need to understand the target distribution and check whether the values are stored correctly.

We will inspect:

- the first target values;
- the data type;
- unique values;
- basic statistics;
- possible extreme prices.

In [46]:
df[target_column].head(10)

0    5500
1    5350
2    7000
3    3300
4    5200
5    3400
6    5000
7    7300
8    6400
9    6132
Name: priceUSD, dtype: int64

### Target data type

The data type of the target variable is important because regression models require numerical target values.

In [47]:
df[target_column].dtype

dtype('int64')

### Unique target values

Checking unique values helps us verify that the target variable contains valid numerical values.

In [48]:
df[target_column].unique()[:10]

array([5500, 5350, 7000, 3300, 5200, 3400, 5000, 7300, 6400, 6132])

## 9. Missing values

Missing values are common in real-world datasets.

Before training a machine learning model, we need to identify:

- which columns contain missing values;
- how many missing values exist;
- whether missing values are present in numerical or categorical features.

In this notebook, we only inspect missing values.

The strategy for handling them will be defined later during the data cleaning and preprocessing phases.

In [49]:
missing_values = df.isna().sum()

missing_values[missing_values > 0]

volume(cm3)      47
drive_unit     1905
segment        5291
dtype: int64

Raw data can also contain missing-like text values, such as `"NaN"`, `"null"`, `"None"`, or empty strings.

These values may look like regular text, but they usually mean that the value is missing.

In [50]:
missing_like_values = ["NaN", "nan", "NULL", "null", "None", "none", "", " "]

for column in df.columns:
    if df[column].dtype == "object":
        count = df[column].astype(str).str.strip().isin(missing_like_values).sum()

        if count > 0:
            print(f"{column}: {count}")

## 10. Numeric column inspection

Numerical features represent measurable characteristics of the vehicles.

Before using these features in machine learning models, we need to inspect:

- detected data types;
- statistical properties;
- possible unrealistic values;
- extreme values that may require cleaning.

The following numerical columns will be analyzed:

- `priceUSD`;
- `year`;
- `mileage(kilometers)`;
- `volume(cm3)`.

In [51]:
numeric_columns = [
    "priceUSD",
    "year",
    "mileage(kilometers)",
    "volume(cm3)"
]

df[numeric_columns].dtypes

priceUSD                 int64
year                     int64
mileage(kilometers)    float64
volume(cm3)            float64
dtype: object

## 11. Categorical column inspection

Categorical features represent characteristics that describe groups or categories of vehicles.

Examples include:

- vehicle brand;
- vehicle model;
- fuel type;
- transmission type;
- vehicle condition;
- color;
- drivetrain;
- vehicle segment.

These features cannot be directly used by most machine learning algorithms and will require encoding during the preprocessing stage.

In this section, we inspect:

- the number of unique categories;
- the actual values;
- possible inconsistencies in category names.

In [52]:
categorical_columns = [
    "make",
    "model",
    "condition",
    "fuel_type",
    "color",
    "transmission",
    "drive_unit",
    "segment"
]

df[categorical_columns].head()

,make,model,condition,fuel_type,color,transmission,drive_unit,segment
0,mazda,2,with mileage,petrol,burgundy,mechanics,front-wheel drive,B
1,mazda,2,with mileage,petrol,black,mechanics,front-wheel drive,B
2,mazda,2,with mileage,petrol,silver,auto,front-wheel drive,B
3,mazda,2,with mileage,diesel,white,mechanics,front-wheel drive,B
4,mazda,2,with mileage,diesel,gray,mechanics,front-wheel drive,B


### Number of unique categories

The number of unique values in each categorical column helps us understand the complexity of each feature.

Columns with many unique values, such as vehicle models, may require special consideration during preprocessing.

In [53]:
for column in categorical_columns:
    print(column)
    print(df[column].nunique(dropna=False))
    print("-" * 40)

make
96
----------------------------------------
model
1034
----------------------------------------
condition
3
----------------------------------------
fuel_type
3
----------------------------------------
color
13
----------------------------------------
transmission
2
----------------------------------------
drive_unit
5
----------------------------------------
segment
10
----------------------------------------


In [54]:
for column in categorical_columns:
    print(column)
    print(df[column].value_counts(dropna=False).head(10))
    print("-" * 40)

make
make
volkswagen       6861
audi             4030
bmw              4013
opel             3779
renault          3713
mercedes-benz    3541
ford             3078
peugeot          2876
nissan           2233
toyota           2177
Name: count, dtype: int64
----------------------------------------
model
model
passat      2086
5-seriya    1476
a6          1276
golf        1070
astra       1013
e-klass      988
3-seriya     823
vectra       815
laguna       768
mondeo       731
Name: count, dtype: int64
----------------------------------------
condition
condition
with mileage    55278
with damage       512
for parts         454
Name: count, dtype: int64
----------------------------------------
fuel_type
fuel_type
petrol        36405
diesel        19792
electrocar       47
Name: count, dtype: int64
----------------------------------------
color
color
black       12385
silver      10075
blue         8083
gray         5807
white        5292
green        3911
other        3397
red          274

### Category value observations

The inspection of categorical values shows that the categories are generally consistent.

The dataset contains expected vehicle-related categories, including:

- different vehicle manufacturers and models;
- vehicle conditions;
- fuel types;
- transmission types;
- colors;
- drivetrain types;
- vehicle segments.

No obvious inconsistencies such as duplicated category names with different formatting were detected.

Some categorical features contain missing values:

- `drive_unit`;
- `segment`.

These missing values will be handled during the cleaning and preprocessing phases.

The `model` and `make` columns contain a large number of unique categories, which may provide valuable information for price prediction but will require appropriate encoding before model training.

In [55]:
for column in categorical_columns:
    print(column)
    print(df[column].value_counts(dropna=False).head(10))
    print("-" * 40)

make
make
volkswagen       6861
audi             4030
bmw              4013
opel             3779
renault          3713
mercedes-benz    3541
ford             3078
peugeot          2876
nissan           2233
toyota           2177
Name: count, dtype: int64
----------------------------------------
model
model
passat      2086
5-seriya    1476
a6          1276
golf        1070
astra       1013
e-klass      988
3-seriya     823
vectra       815
laguna       768
mondeo       731
Name: count, dtype: int64
----------------------------------------
condition
condition
with mileage    55278
with damage       512
for parts         454
Name: count, dtype: int64
----------------------------------------
fuel_type
fuel_type
petrol        36405
diesel        19792
electrocar       47
Name: count, dtype: int64
----------------------------------------
color
color
black       12385
silver      10075
blue         8083
gray         5807
white        5292
green        3911
other        3397
red          274

## 12. Outlier analysis

Outliers are values that are significantly different from the majority of observations.

In a used car price prediction problem, extreme values can affect the regression model because the model learns from historical examples.

In this section, we inspect possible outliers in:

- vehicle prices;
- production years;
- mileage;
- engine volume.

The goal is to identify potential data quality issues before the cleaning phase.

### Price inspection

Vehicle price is the target variable of this regression problem.

We inspect the minimum and maximum prices to identify possible unrealistic values.

In [56]:
df["priceUSD"].describe()

count     56244.000000
mean       7415.456440
std        8316.959261
min          48.000000
25%        2350.000000
50%        5350.000000
75%        9807.500000
max      235235.000000
Name: priceUSD, dtype: float64

In [57]:
print("Minimum price:", df["priceUSD"].min())
print("Maximum price:", df["priceUSD"].max())

Minimum price: 48
Maximum price: 235235


### Production year inspection

The production year should have realistic values.

Very old or future production years may indicate incorrect data entries.

In [58]:
print("Minimum year:", df["year"].min())
print("Maximum year:", df["year"].max())

Minimum year: 1910
Maximum year: 2019


### Mileage inspection

Mileage is an important factor for used car prices.

Extremely high mileage values may represent data errors or unusual vehicles.

In [59]:
df["mileage(kilometers)"].describe()

count    5.624400e+04
mean     2.443956e+05
std      3.210307e+05
min      0.000000e+00
25%      1.370000e+05
50%      2.285000e+05
75%      3.100000e+05
max      9.999999e+06
Name: mileage(kilometers), dtype: float64

### Engine volume inspection

Engine volume is another important vehicle characteristic.

We check this feature for unrealistic minimum or maximum values.

In [60]:
df["volume(cm3)"].describe()

count    56197.000000
mean      2104.860615
std        959.201633
min        500.000000
25%       1600.000000
50%       1996.000000
75%       2300.000000
max      20000.000000
Name: volume(cm3), dtype: float64

## Final EDA Findings and Cleaning Plan

The dataset contains 56,244 used car records with 12 features.

During exploratory data analysis, several points were identified that need to be
handled before model training:

- Column names should be standardized for easier usage in Python.
- Missing values were detected in some features, especially in volume, drive unit,
  and segment columns.
- Numerical features such as price, year, mileage, and engine volume need to be
  checked for invalid values.
- Outlier analysis showed that some extreme values exist, but they may represent
  valid vehicles and should not be removed without further inspection.
- Categorical features need to be encoded before training regression models.
- Additional features such as car age, mileage per year, and engine volume in
  liters can be created during feature engineering.

After these steps, the cleaned dataset will be prepared for preprocessing and
model training.